# Capstone Project - The Battle of the Neighborhoods
### Applied Data Science Capstone by IBM/Coursera
## Table of contents
* [Introduction: Business Problem](#introduction)
* [Data](#data)
* [Methodology](#methodology)
* [Analysis](#analysis)
* [Results and Discussion](#results)
* [Conclusion](#conclusion)

## Introduction: Business Problem <a name="introduction"></a>
An investor is looking to open a new buffet in Adelaide (which serves all kinds of cusines), but she is unsure about the best location for her new venue and needs input for making the decision. Adelaide is rather busy city famous for its famous tourit attractions. The goal is to set a venue's location to maximize the profit. According to an analysis in the Small Business, the 5 factors for choosing a new restaurant location are:

* **Parking:** Ideally, a new restaurant location should have its own parking lot. If that isn’t an option—for example, in a major city—consider partnering with a hotel in the area that has its own parking options. Many famous restaurants are housed in hotels, and for a good reason. Not only is there parking, but the benefit of foot traffic that is staying right upstairs is incalculable.
* **Accessibility:** There’s a reason that major restaurant chains are often located near highway exits: It makes them accessible for customers. Certain restaurants can get away with food or service that isn't the best simply because their locations are so accessible, like restaurants near the Eiffel Tower or Collisseum. There is plenty of foot traffic in urbanized areas, and restaurants only need to attract customers from the street into their business. Most successful restaurants—other than the truly elite—are easy to find, and you will find them in city centers or unique locations throughout the world.
* **Visibility:** This goes along with accessibility and is very important for new restaurant locations. People have to know the restaurant is there, either in person or on their mobile devices. It is why property prices in downtown districts and developed strips are higher than in other areas. They offer a level of visibility that can bring in a great deal of walk-in business. Consider advertising in search engines and social media to enhance your presence across all forms of media. Make sure to register your restaurant in search engines as the type of food you offer and your price point, as it will be easier to attract the clientele you want when they go to search.
* **Population Base:** Are there enough people in the area to support your business? There need to be enough people who live in or pass through the area regularly to keep you busy. To determine a particular area's population base, you could do a site study. However, these can cost up to 25,000 dollars. Most people looking at their first restaurant don’t have enough money in their budget for a professional survey. A less expensive method to determine the population base of a certain area is to use a pie chart, as well as asking the local chamber of commerce and town office for more information. If you would rather pound the pavement, simply walk around the area where you plan to build. Intuition can place a big role in choosing your site.
* **Safety:** workplace safty is important for the restaurant owner as well as workers, should avoid crime-laden areas in the city.

We will use our data science powers to generate a few most promissing neighborhoods based on this criteria. Advantages of each area will then be clearly expressed so that best possible final location can be chosen by stakeholders.

## Data <a name="data"></a>
Based on definition of our problem, factors that will influence our decission are:
* number of existing restaurants in the neighborhood (any type of restaurant)
* population base in each neighborhood

We decided to use regularly spaced grid of locations, centered around city center, to define our neighborhoods.

Following data sources will be needed to extract/generate the required information:
* centers of candidate areas will be generated algorithmically and approximate addresses of centers of those areas will be obtained using **Google Maps API reverse geocoding**
* number of restaurants and their type and location in every neighborhood will be obtained using **Foursquare API**
* coordinate of Berlin center will be obtained using **Google Maps API geocoding** of well known Berlin location (Alexanderplatz)

The dataset is derived from: 
1. List of Adelaide suburbs: https://www.costlessquotes.com.au/postcode_tool/postcode_list_NSW.php
2. Crime dataset 2020: https://data.gov.au/dataset/ds-sa-860126f7-eeb5-4fbc-be44-069aa0467d11/details?q=crime

### Import the Packages

In [5]:
# Import libraries
import numpy as np # library to handle data in a vectorized manner
import json # library to handle JSON files
import datetime as dt # Datetime

!conda install -c conda-forge geopy --yes 
from geopy.geocoders import Nominatim # convert an address into latitude and longitude values

import requests # library to handle requests
from pandas.io.json import json_normalize # tranform JSON file into a pandas dataframe

# Matplotlib and associated plotting modules
import matplotlib.pyplot as plt
import matplotlib.cm as cm
import matplotlib.colors as colors

#from bs4 import BeautifulSoup

# Import k-means from clustering stage
from sklearn.cluster import KMeans

!conda install -c conda-forge folium=0.5.0 --yes 
import folium # map rendering library

print('Libraries imported.')

Solving environment: done

## Package Plan ##

  environment location: /opt/conda/envs/Python36

  added / updated specs: 
    - geopy


The following packages will be downloaded:

    package                    |            build
    ---------------------------|-----------------
    geopy-1.21.0               |             py_0          58 KB  conda-forge
    geographiclib-1.50         |             py_0          34 KB  conda-forge
    ------------------------------------------------------------
                                           Total:          92 KB

The following NEW packages will be INSTALLED:

    geographiclib: 1.50-py_0   conda-forge
    geopy:         1.21.0-py_0 conda-forge


geopy-1.21.0         | 58 KB     | ##################################### | 100% 
geographiclib-1.50   | 34 KB     | ##################################### | 100% 
Preparing transaction: done
Verifying transaction: done
Executing transaction: done
Solving environment: done

## Package Plan ##

  en

### 1. Load the data 
We will scrape the data from the source page and tidy it up by using only Adelaide for this project. 
The dataset is derived from: https://en.wikipedia.org/wiki/List_of_Adelaide_suburbs

In [104]:
import pandas as pd
source = pd.read_html('https://en.wikipedia.org/wiki/List_of_Adelaide_suburbs')
df1 = source[0]
print (df1.shape)
df1.head(5)

(433, 7)


Suburb  PostCode               LGA YearEstab.[citation needed]  \
0        Adelaide      5000  City of Adelaide                        1837   
1  North Adelaide      5006  City of Adelaide                        1837   
2         Auldana      5072  City of Burnside                    1847 [7]   
3        Beaumont      5066  City of Burnside                        1870   
4     Beulah Park      5067  City of Burnside                    1941[10]   

  Dist.[4](km)[citation needed]  Area(ha)[citation needed]  \
0                             -                     1005.0   
1                           0.5                      420.0   
2                             9                      312.0   
3                           5.9                      158.0   
4                             5                       60.0   

  Population[citation needed]  
0                   15,115[5]  
1                   6,950 [6]  
2                      625[8]  
3                    2,557[9]  
4                   1,602[11]

### 2. Data Preprocessing
* Rename the column names
* Drop unneeded columns - only keep Suberb, PostCode and Population
* Drop NaN in Population - as our project aims to count population factor in
* Data cleaning on the citaition for Population column

In [105]:
df1 = df1.rename(columns = {"Postcode":"PostalCode","Suburb":"Neighborhood","Population[citation needed]":"Population"})
#Note: this does not work
#df1.drop(['LGA', 'YearEstab.[citation needed]','Dist.[4](km)[citation needed]','Area(ha)[citation needed]'], axis = 1)
df1.drop(df1.columns[[2,3,4,5]], axis = 1, inplace = True) 
df1.dropna(inplace=True)
df1.reset_index().drop(df1.columns[0],axis=1,inplace = True)
df1['Population'] = df1['Population'].str.replace(r"\[.*\]","")
df1['Population'] = df1['Population'].str.replace(",","")

In [106]:
df = df1.reset_index(drop=True)

In [107]:
df

Neighborhood  PostCode Population
0             Adelaide      5000      15115
1       North Adelaide      5006      6950 
2              Auldana      5072        625
3             Beaumont      5066       2557
4          Beulah Park      5067       1602
5             Burnside      5066       2930
6              Dulwich      5065       1678
7             Eastwood      5063        764
8             Erindale      5066       1186
9            Frewville      5063        874
10             Glenelg      5045       3349
11         Glen Osmond      5064       1990
12            Glenside      5065       2422
13            Glenunga      5064       2004
14      Hazelwood Park      5066       1874
15  Kensington Gardens      5068       2456
16     Kensington Park      5068       2498
17            Leabrook      5068       1429
18     Leawood Gardens      5150       2375
19         Linden Park      5065       1910
20              Magill      5072       8229
21        Mount Osmond      5064       2375
22           Rose Park      5067       1293
23        Rosslyn Park      5072       1460
24                Skye      5072        278
25          St Georges      5064       1253
26           Stonyfell      5066       1326
27      Toorak Gardens      5065       2521
28             Tusmore      5065       1570
29     Waterfall Gully      5066       2522
..                 ...       ...        ...
35              Newton      5074       4451
36            Paradise      5075       6557
37           Rostrevor      5073       7366
38            Tranmere      5073       5073
39         Albert Park      5014       1724
40     Allenby Gardens      5009       1868
41          Athol Park      5012       2084
42            Beverley      5009       1498
43              Bowden      5007        911
44            Brompton      5007       3537
45          Cheltenham      5014       2175
46             Croydon      5008       1238
47          Devon Park      5008        888
48              Findon      5023       6205
49       Flinders Park      5025       4880
50      Fulham Gardens      5024       5942
51              Grange      5022       5916
52        Henley Beach      5022       5562
53  Henley Beach South      5022       2289
54         Kidman Park      5025       3317
55       Port Adelaide      5015       1293
56   Woodville Gardens      5012       2383
57           Salisbury      5108       7551
58      Salisbury East      5109       8850
59   Salisbury Heights      5109       4237
60        Golden Grove      5125       9664
61           Greenwith      5125       9075
62             Modbury      5092       4924
63   Salisbury Heights      5109       4237
64         Walkerville      5081       2370

[65 rows x 3 columns]

### Conclusion 1: Population Base
Adelaide (15115) and Golden Grove (9664) have the most populations. Investors should primarily consider opening their restaurants in these two suburbs.

## Methodology <a name="methodology"></a>

### 1. Adding the latitude and longtitude

In [2]:
!conda install -c conda-forge geocoder --yes
print("Installation Done!")
import geocoder # import geocoder
print("Geo Coder imported!")

Solving environment: done

## Package Plan ##

  environment location: /opt/conda/envs/Python36

  added / updated specs: 
    - geocoder


The following packages will be downloaded:

    package                    |            build
    ---------------------------|-----------------
    certifi-2019.11.28         |   py36h9f0ad1d_1         149 KB  conda-forge
    python_abi-3.6             |          1_cp36m           4 KB  conda-forge
    geocoder-1.38.1            |             py_1          53 KB  conda-forge
    ratelim-0.1.6              |             py_2           6 KB  conda-forge
    ca-certificates-2019.11.28 |       hecc5488_0         145 KB  conda-forge
    openssl-1.1.1d             |       h516909a_0         2.1 MB  conda-forge
    ------------------------------------------------------------
                                           Total:         2.4 MB

The following NEW packages will be INSTALLED:

    geocoder:        1.38.1-py_1       conda-forge
    python_abi:    

In [108]:
def get_geocoder(postal_code_from_df):
     # initialize your variable to None
    lat_lng_coords = None
     # loop until you get the coordinates
    while(lat_lng_coords is None):
        g = geocoder.arcgis('{}, Adelaide, Australia'.format(postal_code_from_df))
        lat_lng_coords = g.latlng
        latitude = lat_lng_coords[0]
        longitude = lat_lng_coords[1]
    return latitude,longitude

In [109]:
df['latitude'], df['longitude'] = zip(*df['PostCode'].apply(get_geocoder))
df.head()

Neighborhood  PostCode Population   latitude   longitude
0        Adelaide      5000      15115 -34.925650  138.599907
1  North Adelaide      5006      6950  -34.909165  138.594803
2         Auldana      5072        625 -34.910540  138.697384
3        Beaumont      5066       2557 -34.943867  138.670460
4     Beulah Park      5067       1602 -34.924875  138.630063

### 2. Explore the neighborhoods in Adelaide

In [110]:
address = 'Adelaide, South Australia'

geolocator = Nominatim(user_agent="Adelaide_SA")
location = geolocator.geocode(address)
latitude = location.latitude
longitude = location.longitude
print('The geograpical coordinates of Adelaide, NSW are {}, {}.'.format(latitude, longitude))

The geograpical coordinates of Adelaide, NSW are -34.9281805, 138.5999312.


In [112]:
map_adelaide = folium.Map(location=[latitude, longitude], zoom_start=11)

for lat, long, label in zip(df['latitude'], df['longitude'], df['PostCode']):
    label = folium.Popup(label,parse_html=True)
    folium.CircleMarker(
        [lat, long],
        radius=5,
        popup=label,
        color='blue',
        fill=True,
        fill_color='#3186cc',
        fill_opacity=0.7,
        parse_html=False).add_to(map_adelaide)
    
map_adelaide

### Foursquare API

In [113]:
CLIENT_ID = '3JDU52VZBVVBWEF3C2Z1GNCA1K4OW1LUGBJCI3PV3JAYZFVV' # your Foursquare ID
CLIENT_SECRET = 'LNWAQOGNJFMKZY3DONHTR2FB5SBGBLLLDUETQN2P5WFI2U2P' # your Foursquare Secret
VERSION = '20200304' # Foursquare API version

print('Your credentails:')
print('CLIENT_ID: ' + CLIENT_ID)
print('CLIENT_SECRET:' + CLIENT_SECRET)

Your credentails:
CLIENT_ID: 3JDU52VZBVVBWEF3C2Z1GNCA1K4OW1LUGBJCI3PV3JAYZFVV
CLIENT_SECRET:LNWAQOGNJFMKZY3DONHTR2FB5SBGBLLLDUETQN2P5WFI2U2P


### 3. Let's create a function to explore all the suburbs in Adelaide.

In [114]:
LIMIT = 100 # limit of number of venues returned by Foursquare API
radius = 500 # define radius

In [115]:
def getNearbyVenues(names, latitudes, longitudes, radius=500):
    
    venues_list=[]
    for name, lat, lng in zip(names, latitudes, longitudes):
        print(name)
            
        # create the API request URL
        url = 'https://api.foursquare.com/v2/venues/explore?&client_id={}&client_secret={}&v={}&ll={},{}&radius={}&limit={}'.format(
            CLIENT_ID, 
            CLIENT_SECRET, 
            VERSION, 
            lat, 
            lng, 
            radius, 
            LIMIT)
            
        # make the GET request
        results = requests.get(url).json()["response"]['groups'][0]['items']
        # return only relevant information for each nearby venue
        venues_list.append([(
            name, 
            lat, 
            lng, 
            v['venue']['name'], 
            v['venue']['location']['lat'], 
            v['venue']['location']['lng'],  
            v['venue']['categories'][0]['name']) for v in results])
        nearby_venues = pd.DataFrame([item for venue_list in venues_list for item in venue_list])
        nearby_venues.columns = ['Neighborhood', 
                  'Neighborhood Latitude', 
                  'Neighborhood Longitude', 
                  'Venue', 
                  'Venue Latitude', 
                  'Venue Longitude', 
                  'Venue Category']
    
    return(nearby_venues)

Now write the code to run the above function on each neighborhood and **create a new dataframe called ade_venues**.

In [116]:
ade_venues = getNearbyVenues(names=df['Neighborhood'],
                                   latitudes=df['latitude'],
                                   longitudes=df['longitude']
                                  )

Adelaide
North Adelaide
Auldana
Beaumont
Beulah Park
Burnside
Dulwich
Eastwood
Erindale
Frewville
Glenelg
Glen Osmond
Glenside
Glenunga
Hazelwood Park
Kensington Gardens
Kensington Park
Leabrook
Leawood Gardens
Linden Park
Magill
Mount Osmond
Rose Park
Rosslyn Park
Skye
St Georges
Stonyfell
Toorak Gardens
Tusmore
Waterfall Gully
Wattle Park
Athelstone
Campbelltown
Hectorville
Magill
Newton
Paradise
Rostrevor
Tranmere
Albert Park
Allenby Gardens
Athol Park
Beverley
Bowden
Brompton
Cheltenham
Croydon
Devon Park
Findon
Flinders Park
Fulham Gardens
Grange
Henley Beach
Henley Beach South
Kidman Park
Port Adelaide
Woodville Gardens
Salisbury
Salisbury East
Salisbury Heights
Golden Grove
Greenwith
Modbury
Salisbury Heights
Walkerville


In [117]:
ade_venues

Neighborhood  Neighborhood Latitude  Neighborhood Longitude  \
0             Adelaide             -34.925650              138.599907   
1             Adelaide             -34.925650              138.599907   
2             Adelaide             -34.925650              138.599907   
3             Adelaide             -34.925650              138.599907   
4             Adelaide             -34.925650              138.599907   
5             Adelaide             -34.925650              138.599907   
6             Adelaide             -34.925650              138.599907   
7             Adelaide             -34.925650              138.599907   
8             Adelaide             -34.925650              138.599907   
9             Adelaide             -34.925650              138.599907   
10            Adelaide             -34.925650              138.599907   
11            Adelaide             -34.925650              138.599907   
12            Adelaide             -34.925650              138.599907   
13            Adelaide             -34.925650              138.599907   
14            Adelaide             -34.925650              138.599907   
15            Adelaide             -34.925650              138.599907   
16            Adelaide             -34.925650              138.599907   
17            Adelaide             -34.925650              138.599907   
18            Adelaide             -34.925650              138.599907   
19            Adelaide             -34.925650              138.599907   
20            Adelaide             -34.925650              138.599907   
21            Adelaide             -34.925650              138.599907   
22            Adelaide             -34.925650              138.599907   
23            Adelaide             -34.925650              138.599907   
24            Adelaide             -34.925650              138.599907   
25            Adelaide             -34.925650              138.599907   
26            Adelaide             -34.925650              138.599907   
27            Adelaide             -34.925650              138.599907   
28            Adelaide             -34.925650              138.599907   
29            Adelaide             -34.925650              138.599907   
..                 ...                    ...                     ...   
432      Port Adelaide             -34.844826              138.507775   
433      Port Adelaide             -34.844826              138.507775   
434  Woodville Gardens             -34.862680              138.545039   
435  Woodville Gardens             -34.862680              138.545039   
436  Woodville Gardens             -34.862680              138.545039   
437  Woodville Gardens             -34.862680              138.545039   
438          Salisbury             -34.755905              138.621376   
439     Salisbury East             -34.775105              138.670498   
440     Salisbury East             -34.775105              138.670498   
441     Salisbury East             -34.775105              138.670498   
442     Salisbury East             -34.775105              138.670498   
443     Salisbury East             -34.775105              138.670498   
444  Salisbury Heights             -34.775105              138.670498   
445  Salisbury Heights             -34.775105              138.670498   
446  Salisbury Heights             -34.775105              138.670498   
447  Salisbury Heights             -34.775105              138.670498   
448  Salisbury Heights             -34.775105              138.670498   
449       Golden Grove             -34.768385              138.727487   
450          Greenwith             -34.768385              138.727487   
451            Modbury             -34.822183              138.682945   
452            Modbury             -34.822183              138.682945   
453  Salisbury Heights             -34.775105              138.670498   
454  Salisbury Heights             -34.775105              138.67049

In [118]:
ade_venues.shape

(462, 7)

In [119]:
print('There are {} uniques categories.'.format(len(ade_venues['Venue Category'].unique())))

There are 111 uniques categories.


### 4. Analyze Each Neighborhood

In [120]:
ade_onehot = pd.get_dummies(ade_venues[['Venue Category']], prefix="", prefix_sep="")

# add neighborhood column back to dataframe
ade_onehot['Neighborhood'] = ade_venues['Neighborhood'] 

# move neighborhood column to the first column
fixed_columns = [ade_onehot.columns[-1]] + list(ade_onehot.columns[:-1])
ade_onehot = ade_onehot[fixed_columns]
ade_onehot.head()

Neighborhood  Afghan Restaurant  Airport Lounge  American Restaurant  \
0     Adelaide                  0               0                    0   
1     Adelaide                  0               0                    0   
2     Adelaide                  0               0                    0   
3     Adelaide                  0               0                    0   
4     Adelaide                  0               0                    0   

   Argentinian Restaurant  Asian Restaurant  Athletics & Sports  \
0                       0                 0                   0   
1                       0                 0                   0   
2                       0                 0                   0   
3                       0                 0                   0   
4                       0                 0                   0   

   Australian Restaurant  BBQ Joint  Bagel Shop  ...  Thai Restaurant  \
0                      0          0           0  ...                0   
1                      0          0           0  ...                0   
2                      0          0           0  ...                0   
3                      1          0           0  ...                0   
4                      0          0           0  ...                0   

   Theater  Thrift / Vintage Store  Trail  Train Station  Tram Station  \
0        0                       0      0              0             0   
1        0                       0      0              0             0   
2        0                       0      0              0             0   
3        0                       0      0              0             0   
4        0                       0      0              0             0   

   Tunnel  Vegetarian / Vegan Restaurant  Vietnamese Restaurant  Wine Bar  
0       0                              0                      0         1  
1       0                              0                      0         0  
2       0                              0                      0         0  
3       0                              0                      0         0  
4       0                              0                      0         0  

[5 rows x 112 columns]

Next, let's group rows by neighborhood and by taking the **mean of the frequency of occurrence of each category**

In [121]:
ade_grouped = ade_onehot.groupby('Neighborhood').mean().reset_index()
ade_grouped

Neighborhood  Afghan Restaurant  Airport Lounge  \
0             Adelaide              0.010        0.000000   
1          Albert Park              0.000        0.000000   
2      Allenby Gardens              0.000        0.000000   
3           Athelstone              0.000        0.000000   
4           Athol Park              0.000        0.000000   
5             Beaumont              0.000        0.000000   
6          Beulah Park              0.000        0.000000   
7             Beverley              0.000        0.000000   
8               Bowden              0.125        0.000000   
9             Brompton              0.125        0.000000   
10            Burnside              0.000        0.000000   
11        Campbelltown              0.000        0.000000   
12          Cheltenham              0.000        0.000000   
13             Croydon              0.000        0.000000   
14          Devon Park              0.000        0.000000   
15             Dulwich              0.000        0.000000   
16            Eastwood              0.000        0.000000   
17            Erindale              0.000        0.000000   
18              Findon              0.000        0.000000   
19           Frewville              0.000        0.000000   
20      Fulham Gardens              0.000        0.000000   
21         Glen Osmond              0.000        0.000000   
22             Glenelg              0.000        0.000000   
23            Glenside              0.000        0.000000   
24            Glenunga              0.000        0.000000   
25        Golden Grove              0.000        0.000000   
26              Grange              0.000        0.000000   
27           Greenwith              0.000        0.000000   
28      Hazelwood Park              0.000        0.000000   
29         Hectorville              0.000        0.000000   
30        Henley Beach              0.000        0.000000   
31  Henley Beach South              0.000        0.000000   
32  Kensington Gardens              0.000        0.000000   
33     Kensington Park              0.000        0.000000   
34            Leabrook              0.000        0.000000   
35     Leawood Gardens              0.000        0.000000   
36         Linden Park              0.000        0.000000   
37             Modbury              0.000        0.000000   
38        Mount Osmond              0.000        0.000000   
39              Newton              0.000        0.000000   
40      North Adelaide              0.000        0.000000   
41            Paradise              0.000        0.333333   
42       Port Adelaide              0.000        0.000000   
43           Rose Park              0.000        0.000000   
44           Rostrevor              0.000        0.000000   
45           Salisbury              0.000        0.000000   
46      Salisbury East              0.000        0.000000   
47   Salisbury Heights              0.000        0.000000   
48          St Georges              0.000        0.000000   
49           Stonyfell              0.000        0.000000   
50      Toorak Gardens              0.000        0.000000   
51            Tranmere              0.000        0.000000   
52             Tusmore              0.000        0.000000   
53         Walkerville              0.000        0.000000   
54     Waterfall Gully              0.000        0.000000   
55         Wattle Park              0.000        0.000000   
56   Woodville Gardens              0.000        0.000000   

    American Restaurant  Argentinian Restaurant  Asian Restaurant  \
0                  0.01                    0.01          0.030000   
1                  0.00                    0.00          0.000000   
2                  0.00                    0.00          0.250000   
3                  0.00                    0.00          0.000000   
4                  0.00                    0.00          0.000000   
5                  0.00                    0.00          

### Let's print each neighborhood along with the top 5 most common venues

In [122]:
num_top_venues = 5

for hood in ade_grouped['Neighborhood']:
    print("----"+hood+"----")
    temp = ade_grouped[ade_grouped['Neighborhood'] == hood].T.reset_index()
    temp.columns = ['venue','freq']
    temp = temp.iloc[1:]
    temp['freq'] = temp['freq'].astype(float)
    temp = temp.round({'freq': 2})
    print(temp.sort_values('freq', ascending=False).reset_index(drop=True).head(num_top_venues))
    print('\n')

----Adelaide----
              venue  freq
0              Café  0.11
1               Bar  0.09
2       Coffee Shop  0.08
3  Sculpture Garden  0.04
4             Hotel  0.04


----Albert Park----
                   venue  freq
0  Vietnamese Restaurant  0.25
1            IT Services  0.25
2          Grocery Store  0.25
3       Malay Restaurant  0.25
4      Afghan Restaurant  0.00


----Allenby Gardens----
                venue  freq
0  Italian Restaurant  0.25
1    Asian Restaurant  0.25
2                Café  0.25
3           Pet Store  0.25
4               Motel  0.00


----Athelstone----
               venue  freq
0        Pizza Place  0.25
1      Grocery Store  0.25
2             Bakery  0.25
3      Shopping Mall  0.25
4  Afghan Restaurant  0.00


----Athol Park----
                   venue  freq
0  Vietnamese Restaurant  0.50
1              Pet Store  0.25
2          Grocery Store  0.25
3      Afghan Restaurant  0.00
4      Mobile Phone Shop  0.00


----Beaumont----
               v

### Let's put that into a pandas dataframe
First, let's write a function to **sort the venues in descending order**.

In [123]:
def return_most_common_venues(row, num_top_venues):
    row_categories = row.iloc[1:]
    row_categories_sorted = row_categories.sort_values(ascending=False)
    
    return row_categories_sorted.index.values[0:num_top_venues]

Now let's create the new dataframe and display the **top 10 venues for each neighborhood.**

In [132]:
num_top_venues = 10
indicators = ['st', 'nd', 'rd']

# create columns according to number of top venues
columns = ['Neighborhood']
for ind in np.arange(num_top_venues):
    try:
        columns.append('{}{} Most Common Venue'.format(ind+1, indicators[ind]))
    except:
        columns.append('{}th Most Common Venue'.format(ind+1))

# create a new dataframe
neighborhoods_venues_sorted = pd.DataFrame(columns=columns)
neighborhoods_venues_sorted['Neighborhood'] = ade_grouped['Neighborhood']

for ind in np.arange(ade_grouped.shape[0]):
    neighborhoods_venues_sorted.iloc[ind, 1:] = return_most_common_venues(ade_grouped.iloc[ind, :], num_top_venues)

neighborhoods_venues_sorted.head()

Neighborhood  1st Most Common Venue  2nd Most Common Venue  \
0         Adelaide                   Café                    Bar   
1      Albert Park            IT Services  Vietnamese Restaurant   
2  Allenby Gardens              Pet Store     Italian Restaurant   
3       Athelstone            Pizza Place          Grocery Store   
4       Athol Park  Vietnamese Restaurant              Pet Store   

  3rd Most Common Venue 4th Most Common Venue 5th Most Common Venue  \
0           Coffee Shop                 Hotel      Sculpture Garden   
1         Grocery Store      Malay Restaurant  Fast Food Restaurant   
2      Asian Restaurant                  Café              Wine Bar   
3                Bakery         Shopping Mall     Electronics Store   
4         Grocery Store        Farmers Market           Coffee Shop   

     6th Most Common Venue    7th Most Common Venue    8th Most Common Venue  \
0                 Wine Bar      Japanese Restaurant       Italian Restaurant   
1  Comfort Food Restaurant        Convenience Store             Cupcake Shop   
2           Farmers Market  Comfort Food Restaurant        Convenience Store   
3             Cocktail Bar              Coffee Shop  Comfort Food Restaurant   
4  Comfort Food Restaurant        Convenience Store             Cupcake Shop   

  9th Most Common Venue 10th Most Common Venue  
0      Asian Restaurant  Vietnamese Restaurant  
1         Deli / Bodega       Department Store  
2          Cupcake Shop          Deli / Bodega  
3     Convenience Store           Cupcake Shop  
4         Deli / Bodega       Department Store

### 5. Clustering the Neighborhoods
Run k-means to cluster the neighborhood into **5 clusters.**

In [133]:
kclusters = 5

ade_grouped_clustering = ade_grouped.drop('Neighborhood', 1)

# run k-means clustering
kmeans = KMeans(n_clusters=kclusters, random_state=0).fit(ade_grouped_clustering)

# check cluster labels generated for each row in the dataframe
kmeans.labels_[0:10]

array([0, 0, 0, 0, 0, 2, 0, 0, 0, 0], dtype=int32)

Now, let's create a new dataframe that includes the **cluster** as well as the **top 10 venues for each neighborhood.**

In [134]:
neighborhoods_venues_sorted.insert(0, 'Cluster_Labels', kmeans.labels_)
neighborhoods_venues_sorted=neighborhoods_venues_sorted.set_index('Neighborhood')

df_right = pd.merge(df, neighborhoods_venues_sorted, on='Neighborhood', how='right')
df_right.head(2)

Neighborhood  PostCode Population   latitude   longitude  Cluster_Labels  \
0        Adelaide      5000      15115 -34.925650  138.599907               0   
1  North Adelaide      5006      6950  -34.909165  138.594803               0   

  1st Most Common Venue 2nd Most Common Venue 3rd Most Common Venue  \
0                  Café                   Bar           Coffee Shop   
1                   Pub          Burger Joint    Italian Restaurant   

  4th Most Common Venue 5th Most Common Venue 6th Most Common Venue  \
0                 Hotel      Sculpture Garden              Wine Bar   
1       Thai Restaurant      Asian Restaurant                  Café   

  7th Most Common Venue 8th Most Common Venue 9th Most Common Venue  \
0   Japanese Restaurant    Italian Restaurant      Asian Restaurant   
1           Coffee Shop                   Gym            Restaurant   

  10th Most Common Venue  
0  Vietnamese Restaurant  
1                   Park

In [135]:
import matplotlib.cm as cm
import matplotlib.colors as colors
# create map
map_clusters = folium.Map(location=[latitude, longitude], zoom_start=11)

# set color scheme for the clusters
x = np.arange(kclusters)
ys = [i + x + (i*x)**2 for i in range(kclusters)]
colors_array = cm.rainbow(np.linspace(0, 1, len(ys)))
rainbow = [colors.rgb2hex(i) for i in colors_array]

# add markers to the map
markers_colors = []
for lat, lon, poi, cluster in zip(df_right['latitude'], df_right['longitude'], df_right['Neighborhood'], df_right['Cluster_Labels']):
    label = folium.Popup(str(poi) + ' Cluster ' + str(cluster), parse_html=True)
    folium.CircleMarker(
        [lat, lon],
        radius=5,
        popup=label,
        color=rainbow[cluster-1],
        fill=True,
        fill_color=rainbow[cluster-1],
        fill_opacity=0.7).add_to(map_clusters)
       
map_clusters

In [136]:
is_0 =  df_right['Cluster_Labels']==0
cluster0 = df_right[is_0]
cluster0

Neighborhood  PostCode Population   latitude   longitude  \
0            Adelaide      5000      15115 -34.925650  138.599907   
1      North Adelaide      5006      6950  -34.909165  138.594803   
3         Beulah Park      5067       1602 -34.924875  138.630063   
5             Dulwich      5065       1678 -34.938560  138.638117   
6            Eastwood      5063        764 -34.952412  138.622420   
8           Frewville      5063        874 -34.952412  138.622420   
9             Glenelg      5045       3349 -34.977801  138.522150   
11           Glenside      5065       2422 -34.938560  138.638117   
17    Leawood Gardens      5150       2375 -34.970402  138.676470   
18        Linden Park      5065       1910 -34.938560  138.638117   
20          Rose Park      5067       1293 -34.924875  138.630063   
23     Toorak Gardens      5065       2521 -34.938560  138.638117   
24            Tusmore      5065       1570 -34.938560  138.638117   
27         Athelstone      5076       9289 -34.871305  138.706915   
28       Campbelltown      5074       7003 -34.880490  138.669605   
30             Newton      5074       4451 -34.880490  138.669605   
31           Paradise      5075       6557 -34.866850  138.679196   
34        Albert Park      5014       1724 -34.868270  138.516056   
35    Allenby Gardens      5009       1868 -34.894938  138.549065   
36         Athol Park      5012       2084 -34.862680  138.545039   
37           Beverley      5009       1498 -34.894938  138.549065   
38             Bowden      5007        911 -34.901890  138.570164   
39           Brompton      5007       3537 -34.901890  138.570164   
40         Cheltenham      5014       2175 -34.868270  138.516056   
43             Findon      5023       6205 -34.893035  138.518541   
44     Fulham Gardens      5024       5942 -34.932697  138.509790   
48      Port Adelaide      5015       1293 -34.844826  138.507775   
49  Woodville Gardens      5012       2383 -34.862680  138.545039   
50          Salisbury      5108       7551 -34.755905  138.621376   
56            Modbury      5092       4924 -34.822183  138.682945   
57        Walkerville      5081       2370 -34.894405  138.611607   

    Cluster_Labels  1st Most Common Venue  2nd Most Common Venue  \
0                0                   Café                    Bar   
1                0                    Pub           Burger Joint   
3                0           Liquor Store                    Pub   
5                0     Chinese Restaurant                   Café   
6                0  Vietnamese Restaurant       Asian Restaurant   
8                0  Vietnamese Restaurant       Asian Restaurant   
9                0   Gym / Fitness Center        Thai Restaurant   
11               0     Chinese Restaurant                   Café   
17               0                 Tunnel                  Trail   
18               0     Chinese Restaurant                   Café   
20               0           Liquor Store                    Pub   
23               0     Chinese Restaurant                   Café   
24               0     Chinese Restaurant                   Café   
27               0            Pizza Place          Grocery Store   
28               0               Pharmacy              BBQ Joint   
30               0               Pharmacy              BBQ Joint   
31               0         Airport Lounge            Gas Station   
34               0            IT Services  Vietnamese Restaurant   
35               0              Pet Store     Italian Restaurant   
36               0  Vietnamese Restaurant              Pet Store   
37               0              Pet Store     Italian Restaurant   
38               0      Indian Restaurant                    Pub   
39               0      Indian Restaurant                    Pub   
40               0            IT Services  Vietnamese Restaurant   
43               0  Vietnamese Restaurant      Recreation Center   
44               0           Ten

If we use the same method on filtering other clusters, we could find: 


**Cluster0:** restaurants and similiar business stalls.    
**Cluster1:** clothing store and tons of restaurants.  
**Cluster2:** Scenic Lookout and some food stalls/Delis  
**Cluster3:** Cafe and snack places  
**Cluster4:** Beer Garden, Bus station, motel  
### Conclusion 2:  
Investors should potentially consider choosing the neighborhoods to avoid fierce competitiveness. 
Based on our observation, **Cluster2 could be the ideal place to launch a restaurant business**, as it is primarily famous by its Scenic Lookout, hence will have enough customer base (i.e.,tourists) to support the business in the long run.

In [137]:
df_right

Neighborhood  PostCode Population   latitude   longitude  \
0             Adelaide      5000      15115 -34.925650  138.599907   
1       North Adelaide      5006      6950  -34.909165  138.594803   
2             Beaumont      5066       2557 -34.943867  138.670460   
3          Beulah Park      5067       1602 -34.924875  138.630063   
4             Burnside      5066       2930 -34.943867  138.670460   
5              Dulwich      5065       1678 -34.938560  138.638117   
6             Eastwood      5063        764 -34.952412  138.622420   
7             Erindale      5066       1186 -34.943867  138.670460   
8            Frewville      5063        874 -34.952412  138.622420   
9              Glenelg      5045       3349 -34.977801  138.522150   
10         Glen Osmond      5064       1990 -34.959290  138.647426   
11            Glenside      5065       2422 -34.938560  138.638117   
12            Glenunga      5064       2004 -34.959290  138.647426   
13      Hazelwood Park      5066       1874 -34.943867  138.670460   
14  Kensington Gardens      5068       2456 -34.920855  138.654085   
15     Kensington Park      5068       2498 -34.920855  138.654085   
16            Leabrook      5068       1429 -34.920855  138.654085   
17     Leawood Gardens      5150       2375 -34.970402  138.676470   
18         Linden Park      5065       1910 -34.938560  138.638117   
19        Mount Osmond      5064       2375 -34.959290  138.647426   
20           Rose Park      5067       1293 -34.924875  138.630063   
21          St Georges      5064       1253 -34.959290  138.647426   
22           Stonyfell      5066       1326 -34.943867  138.670460   
23      Toorak Gardens      5065       2521 -34.938560  138.638117   
24             Tusmore      5065       1570 -34.938560  138.638117   
25     Waterfall Gully      5066       2522 -34.943867  138.670460   
26         Wattle Park      5066       1830 -34.943867  138.670460   
27          Athelstone      5076       9289 -34.871305  138.706915   
28        Campbelltown      5074       7003 -34.880490  138.669605   
29         Hectorville      5073       3477 -34.899590  138.683160   
30              Newton      5074       4451 -34.880490  138.669605   
31            Paradise      5075       6557 -34.866850  138.679196   
32           Rostrevor      5073       7366 -34.899590  138.683160   
33            Tranmere      5073       5073 -34.899590  138.683160   
34         Albert Park      5014       1724 -34.868270  138.516056   
35     Allenby Gardens      5009       1868 -34.894938  138.549065   
36          Athol Park      5012       2084 -34.862680  138.545039   
37            Beverley      5009       1498 -34.894938  138.549065   
38              Bowden      5007        911 -34.901890  138.570164   
39            Brompton      5007       3537 -34.901890  138.570164   
40          Cheltenham      5014       2175 -34.868270  138.516056   
41             Croydon      5008       1238 -34.888220  138.567377   
42          Devon Park      5008        888 -34.888220  138.567377   
43              Findon      5023       6205 -34.893035  138.518541   
44      Fulham Gardens      5024       5942 -34.932697  138.509790   
45              Grange      5022       5916 -34.909655  138.495500   
46        Henley Beach      5022       5562 -34.909655  138.495500   
47  Henley Beach South      5022       2289 -34.909655  138.495500   
48       Port Adelaide      5015       1293 -34.844826  138.507775   
49   Woodville Gardens      5012       2383 -34.862680  138.545039   
50           Salisbury      5108       7551 -34.755905  138.621376   
51      Salisbury East      5109       8850 -34.775105  138.670498   
52   Salisbury Heights      5109       4237 -34.775105  138.670498   
53   Salisbury Heights      5109       4237 -34.775105  138.670498   
54        Golden Grove      5125       9664 -34.768385  138.727487   
55           Greenwith      5125       9075 -34.768385  138.727487   
56             Modbu

## Results: 
After a closer look into factors: population base and visibility. Using population data, we can conclude that Adelaide and Golden Grove have the most populations out of all neighborhoods (limited to the data we could access with). Investors should primarily consider opening their restaurants in these two suburbs.
Another factor, which is visibility, it's safe to conclude neighborhoods in cluster2 should be invested, as it is filled with Scenic Lookout and less competitive to opening a restaurant.  

In [138]:
is_2 =  df_right['Cluster_Labels']==2
cluster2 = df_right[is_2]
cluster2

Neighborhood  PostCode Population   latitude  longitude  \
2          Beaumont      5066       2557 -34.943867  138.67046   
4          Burnside      5066       2930 -34.943867  138.67046   
7          Erindale      5066       1186 -34.943867  138.67046   
13   Hazelwood Park      5066       1874 -34.943867  138.67046   
22        Stonyfell      5066       1326 -34.943867  138.67046   
25  Waterfall Gully      5066       2522 -34.943867  138.67046   
26      Wattle Park      5066       1830 -34.943867  138.67046   

    Cluster_Labels 1st Most Common Venue 2nd Most Common Venue  \
2                2        Scenic Lookout      Department Store   
4                2        Scenic Lookout      Department Store   
7                2        Scenic Lookout      Department Store   
13               2        Scenic Lookout      Department Store   
22               2        Scenic Lookout      Department Store   
25               2        Scenic Lookout      Department Store   
26               2        Scenic Lookout      Department Store   

   3rd Most Common Venue 4th Most Common Venue 5th Most Common Venue  \
2               Wine Bar        Clothing Store           Coffee Shop   
4               Wine Bar        Clothing Store           Coffee Shop   
7               Wine Bar        Clothing Store           Coffee Shop   
13              Wine Bar        Clothing Store           Coffee Shop   
22              Wine Bar        Clothing Store           Coffee Shop   
25              Wine Bar        Clothing Store           Coffee Shop   
26              Wine Bar        Clothing Store           Coffee Shop   

      6th Most Common Venue 7th Most Common Venue 8th Most Common Venue  \
2   Comfort Food Restaurant     Convenience Store          Cupcake Shop   
4   Comfort Food Restaurant     Convenience Store          Cupcake Shop   
7   Comfort Food Restaurant     Convenience Store          Cupcake Shop   
13  Comfort Food Restaurant     Convenience Store          Cupcake Shop   
22  Comfort Food Restaurant     Convenience Store          Cupcake Shop   
25  Comfort Food Restaurant     Convenience Store          Cupcake Shop   
26  Comfort Food Restaurant     Convenience Store          Cupcake Shop   

   9th Most Common Venue 10th Most Common Venue  
2          Deli / Bodega           Dessert Shop  
4          Deli / Bodega           Dessert Shop  
7          Deli / Bodega           Dessert Shop  
13         Deli / Bodega           Dessert Shop  
22         Deli / Bodega           Dessert Shop  
25         Deli / Bodega           Dessert Shop  
26         Deli / Bodega           Dessert Shop

## Discussion: 
This project aims to find the best location for the clients who want to open a restaurant in Adelaide, capital of South Australia. As there are many other factors that we mentioned above could affect the business performance for restaurants, in future studies, we could dive into other factors and give a more comprehensive look into our decisions, such as Safety and Affordability.

We may also conduct a competitor analysis, so we can get a grasp on what our peers doing and how well they are.  

Parking: This is a rather important factor that is not addressed in our analysis. While we have public parking space data, there are many private parkings in San Francisco and we don't have the data. The next step is to obtain a distribution of these parking spaces. Or one could extract the parking information in realtime by using the ParkWhiz API calls.

## Conclusion: 
In this capstone project, we address the business problem of finding a good location in Adelaide for opening a new restaurant. We identify the most important factors that could impact the choice. Using the list of adelaide suburb datasets, we are able to extract the important information on population base. 
By employing Foursquare API recommendation, we could cluster the neighborhoods into 5 clusters and we have found cluster 2 is the most invest-worthy cluster, within it we have: Beaumont, Burnside, Stonyfell, Wattle Park, Erindale, Waterfall Gully, and Hazelwood Park.	
In future studies, we could adopt machine learning alogorithm to help our clients make prediction on cost-revenue analysis. 